packages/editor/src/state/callbacks.ts

```diff
@@ -84,34 +86,50 @@ const _compileDiagram = async (
     }));
   };
 
-  const id = optimizer.run({
-    domain,
-    style,
-    substance,
-    variation,
-    onUpdate,
-    onError: (error) => {
-      toast.dismiss(compiling);
-      set(diagramState, (state: Diagram) => ({ ...state, error }));
+  const onError = (error: any) => {
+    toast.dismiss(compiling);
+    toast.error(error.message);
+    set(diagramWorkerState, {
+      ...diagramWorkerState,
+      optimizing: false,
+    });
+  };
+
+  // ugly `then` chain allows for one catch at the end
```

Comment:
I see that a lot of your usage code uses `.then`/`.catch` instead of `async`/`await`; is this necessary? Is there no way to write this using normal `async`/`await` code?

Reply:
In some parts the `thens` are necessary because I don't want to synchronously wait for the promise to resolve, I just want to attach a callback. In this particular section I was being silly--I wanted to save on catches for every single promise, but really I think you're right that it is more readable to do a bunch of awaits and then `try`/`catch` around that. I'll fix that in a bit.